lib/gruf/errors/handlers.rb

```diff
@@ -0,0 +1,77 @@
+# coding: utf-8
+# Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+module Gruf
+  module Errors
+    module Handlers
+      ##
+      # Will issue a GRPC BadStatus exception, with a code based on the code passed.
+      #
+      # @param [Symbol] error_code The network error code that maps to gRPC status codes
+      # @param [Symbol] app_code The application-specific code for the error
+      # @param [String] message (Optional) A detail message about the error
+      # @param [Hash] metadata (Optional) Any metadata to inject into the trailing metadata for the response
+      #
+      def fail!(error_code, app_code = nil, message = '', metadata = {})
+        e = error
+        e.code = error_code.to_sym
+        e.app_code = app_code ? app_code.to_sym : error.code
+        e.message = message.to_s
+        e.metadata = metadata
+        e.fail!(request.active_call)
+      end
+
+      ##
+      # Add a field error to this endpoint
+      #
+      # @param [Symbol] field_name The name of the field
+      # @param [Symbol] error_code The application error code for the field
+      # @param [String] message A given error message for the field
+      #
+      def add_field_error(field_name, error_code, message = '')
+        error.add_field_error(field_name, error_code, message)
```

Comment:
This makes me feel like `@error` should be an instance variable on the controller instance and this module just exposes convenience methods for interacting with it. Not a blocker, but would be nice to have I think as the controller should be responsible for managing the error state. And then this is more of an `Errors::Helpers` mixin instead of a handler.

Reply:
To do this, I'm going to actually pass that instance variable into the interceptor initializers as well, so interceptors will be able to modify error state. I think this is useful for a few reasons - allowing interceptors to "fail gracefully" and handle multiple interceptors failing but not exiting early; also to allow field-level validation to move upstream without necessarily breaking out of the request cycle prior to the main method call.

Reply:
That's extremely powerful! Good call.